# Análisis de desórdenes mentales en el mundo.

## Introducción

## Justificación
El presente proyecto busca generar un análisis de la evolución de diversos desórdenes mentales en el mundo a lo largos de las últimas décadas. Para este proyecto se hace especial énfasis en la depresión, ya que se cuenta con una base de información extremadamente bien documentada sobre este fenómeno en particular y, por otro lado, este desorden en particular se encuentra como la mayor causa de discapacidad en el mundo, siendo entonces su estudio y comprensión de suma importancia para un mayor entendimiento de sus causas, consecuencias, así como de sus posibles soluciones.

La OMS ha identificado fuertes vínculos entre la depresión y otros trastornos y enfermedades no transmisibles. La depresión aumenta el riesgo de trastornos por uso de sustancias y enfermedades como la diabetes y las enfermedades del corazón; lo contrario también es cierto, lo que significa que las personas con estas otras condiciones tienen un mayor riesgo de depresión.

La depresión también es un factor de riesgo importante para el suicidio, que reclama cientos de miles de vidas cada año.

Es por esta razón que un análisis que nos permita estudiarla resulta sumamente relevante.

## Preguntas clave
- ¿A qué edad empiezan a manifestarse ciertos desórdenes?
- ¿En qué zonas del mundo hay mayor prevalencia de desórdenes mentales?
- ¿Qué tan prevalentes son estos desórdenes en países desarrollados? ¿Y en países en desarrollo?
- ¿Qué tendencia tienen en su prevalencia a lo largo de los años?
- ¿Está relacionada la depresión con el suicidio?

### Instalación de las librerías

In [2]:
#!pip install openpyxl
#!pip install pandas
#!pip install numpy

### Importación de Librerías
- Pandas

In [3]:
import pandas as pd

### Lectura de datos

In [4]:
global_disorders = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='prevalence-by-mental-and-substa')
education_level = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='depression-by-level-of-educatio')
age_groups = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='prevalence-of-depression-by-age')
gender_groups = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='prevalence-of-depression-males-')
depression_and_suicide_rate = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='suicide-rates-vs-prevalence-of-')
affected = pd.read_excel(open('./data/depression.xlsx', 'rb'), sheet_name='number-with-depression-by-count')

### Visualización las dimensiones de las tablas

La uniformidad de los datos es importante, ya que nos permite concatenar fácilmente datos que están relacionados entre sí.

Esto se puede realizar mediante la función <code>shape</code>.

In [5]:
print('global disorders: ', global_disorders.shape)
print('education levels', education_level.shape)
print('age groups: ', age_groups.shape)
print('gender groups: ', gender_groups.shape)
print('depression and suicide rate: ', depression_and_suicide_rate.shape)
print('affected: ', affected.shape)

global disorders:  (6468, 10)
education levels (26, 15)
age groups:  (6468, 13)
gender groups:  (47807, 6)
depression and suicide rate:  (47807, 6)
affected:  (6468, 4)


### Nombre y tipos de datos

La consistencia entre los metadatos de un conjunto de datos permite la correcta integración entre sí.

Esto se puede realizar mediante la función <code>dtypes</code>.

Tabla de índices de desórdenes globales

In [6]:
global_disorders.dtypes

Entity                        object
Code                          object
Year                           int64
Schizophrenia (%)            float64
Bipolar disorder (%)         float64
Eating disorders (%)         float64
Anxiety disorders (%)        float64
Drug use disorders (%)       float64
Depression (%)               float64
Alcohol use disorders (%)    float64
dtype: object

Tabla de índices de depresión por nivel de educación

In [7]:
education_level.dtypes

Entity                                                           object
Code                                                             object
Year                                                              int64
All levels (active) (%)                                         float64
All levels (employed) (%)                                       float64
All levels (total) (%)                                          float64
Below upper secondary (active) (%)                              float64
Below upper secondary (employed) (%)                            float64
Below upper secondary (total) (%)                               float64
Tertiary (active) (%)                                           float64
Tertiary (employed) (%)                                         float64
Tertiary (total) (%)                                            float64
Upper secondary & post-secondary non-tertiary (active) (%)      float64
Upper secondary & post-secondary non-tertiary (employed) (%)    

Tabla de índices de depresión por edad

In [8]:
age_groups.dtypes

Entity                   object
Code                     object
Year                      int64
20-24 years old (%)     float64
10-14 years old (%)     float64
All ages (%)            float64
70+ years old (%)       float64
30-34 years old (%)     float64
15-19 years old (%)     float64
25-29 years old (%)     float64
50-69 years old (%)     float64
Age-standardized (%)    float64
15-49 years old (%)     float64
dtype: object

Tabla de índices de depresión por género

In [9]:
gender_groups.dtypes

Entity                        object
Code                          object
Year                          object
Prevalence in males (%)      float64
Prevalence in females (%)    float64
Population                   float64
dtype: object

Tabla de índices de depresión y suicidio por cada 100,000 individuos

In [10]:
depression_and_suicide_rate.dtypes

Entity                                                       object
Code                                                         object
Year                                                         object
Suicide rate (deaths per 100,000 individuals)               float64
Depressive disorder rates (number suffering per 100,000)    float64
Population                                                  float64
dtype: object

Tabla de población afectada por la depresión

In [11]:
affected.dtypes

Entity                                                                                                        object
Code                                                                                                          object
Year                                                                                                           int64
Prevalence - Depressive disorders - Sex: Both - Age: All Ages (Number) (people suffering from depression)    float64
dtype: object

### Visualización de un conjunto de datos pequeño

Explorar el conjunto de datos nos permite visualizar los posibles valores que adquiere cada columna y de esta forma, realizar un correcto procesamiento de datos.

Esto se puede realizar mediante la función <code>head</code>.

Tabla de índices de desórdenes globales

In [12]:
global_disorders.head()

Entity Code  Year  Schizophrenia (%)  Bipolar disorder (%)  \
0  Afghanistan  AFG  1990           0.160560              0.697779   
1  Afghanistan  AFG  1991           0.160312              0.697961   
2  Afghanistan  AFG  1992           0.160135              0.698107   
3  Afghanistan  AFG  1993           0.160037              0.698257   
4  Afghanistan  AFG  1994           0.160022              0.698469   

   Eating disorders (%)  Anxiety disorders (%)  Drug use disorders (%)  \
0              0.101855               4.828830                1.677082   
1              0.099313               4.829740                1.684746   
2              0.096692               4.831108                1.694334   
3              0.094336               4.830864                1.705320   
4              0.092439               4.829423                1.716069   

   Depression (%)  Alcohol use disorders (%)  
0        4.071831                   0.672404  
1        4.079531                   0.671768  
2        4.088358                   0.670644  
3        4.096190                   0.669738  
4        4.099582                   0.669260

Tabla de índices de depresión por nivel de educación

In [13]:
education_level.head()

Entity Code  Year  All levels (active) (%)  \
0         Austria  AUT  2014                      6.5   
1         Belgium  BEL  2014                      5.0   
2  Czech Republic  CZE  2014                      3.0   
3         Denmark  DNK  2014                      6.7   
4         Estonia  EST  2014                      3.8   

   All levels (employed) (%)  All levels (total) (%)  \
0                        4.7                     7.7   
1                        4.1                     7.1   
2                        2.6                     4.0   
3                        5.7                     8.3   
4                        3.8                     5.1   

   Below upper secondary (active) (%)  Below upper secondary (employed) (%)  \
0                                15.5                                   9.0   
1                                 7.1                                   4.8   
2                                 2.1                                   2.5   
3                                10.4                                   6.5   
4                                 4.7                                   4.7   

   Below upper secondary (total) (%)  Tertiary (active) (%)  \
0                               15.2                    4.3   
1                               11.6                    3.7   
2                                6.0                    1.7   
3                               15.5                    5.7   
4                                6.4                    3.6   

   Tertiary (employed) (%)  Tertiary (total) (%)  \
0                      3.5                   5.5   
1                      3.3                   4.2   
2                      1.7                   2.0   
3                      4.7                   6.7   
4                      3.6                   4.3   

   Upper secondary & post-secondary non-tertiary (active) (%)  \
0                                                5.5            
1                                                5.7            
2                                                3.5            
3                                                7.4            
4                                                3.7            

   Upper secondary & post-secondary non-tertiary (employed) (%)  \
0                                                4.2              
1                                                5.0              
2                                                3.0              
3                                                6.9              
4                                                3.8              

   Upper secondary & post-secondary non-tertiary (total) (%)  
0                                                6.7          
1                                                7.5          
2                                                4.4          
3                                                8.8          
4                                                5.2

Tabla de índices de depresión por edad

In [14]:
age_groups.head()

Entity Code  Year  20-24 years old (%)  10-14 years old (%)  \
0  Afghanistan  AFG  1990             4.417802             1.594676   
1  Afghanistan  AFG  1991             4.433524             1.588356   
2  Afghanistan  AFG  1992             4.453689             1.577980   
3  Afghanistan  AFG  1993             4.464517             1.577201   
4  Afghanistan  AFG  1994             4.462960             1.570846   

   All ages (%)  70+ years old (%)  30-34 years old (%)  15-19 years old (%)  \
0      3.218871           5.202803             5.799034             3.455708   
1      3.203468           5.192849             5.814828             3.451880   
2      3.156559           5.176872             5.829745             3.434982   
3      3.120655           5.167355             5.853060             3.420210   
4      3.082179           5.157549             5.852851             3.425222   

   25-29 years old (%)  50-69 years old (%)  Age-standardized (%)  \
0             5.175856             5.917752              4.071831   
1             5.176729             5.927093              4.079531   
2             5.160249             5.945656              4.088358   
3             5.148767             5.966915              4.096190   
4             5.148227             5.975907              4.099582   

   15-49 years old (%)  
0             4.939766  
1             4.902682  
2             4.837097  
3             4.813657  
4             4.839340

Tabla de índices de depresión por género

In [15]:
gender_groups.head()

Entity Code  Year  Prevalence in males (%)  Prevalence in females (%)  \
0  Afghanistan  AFG  1800                      NaN                        NaN   
1  Afghanistan  AFG  1801                      NaN                        NaN   
2  Afghanistan  AFG  1802                      NaN                        NaN   
3  Afghanistan  AFG  1803                      NaN                        NaN   
4  Afghanistan  AFG  1804                      NaN                        NaN   

   Population  
0   3280000.0  
1   3280000.0  
2   3280000.0  
3   3280000.0  
4   3280000.0

Tabla de índices de depresión y suicidio por cada 100,000 individuos

In [16]:
depression_and_suicide_rate.head()

Entity Code  Year  Suicide rate (deaths per 100,000 individuals)  \
0  Afghanistan  AFG  1800                                            NaN   
1  Afghanistan  AFG  1801                                            NaN   
2  Afghanistan  AFG  1802                                            NaN   
3  Afghanistan  AFG  1803                                            NaN   
4  Afghanistan  AFG  1804                                            NaN   

   Depressive disorder rates (number suffering per 100,000)  Population  
0                                                NaN          3280000.0  
1                                                NaN          3280000.0  
2                                                NaN          3280000.0  
3                                                NaN          3280000.0  
4                                                NaN          3280000.0

Tabla de población afectada por la depresión

In [17]:
affected.head()

Entity Code  Year  \
0  Afghanistan  AFG  1990   
1  Afghanistan  AFG  1991   
2  Afghanistan  AFG  1992   
3  Afghanistan  AFG  1993   
4  Afghanistan  AFG  1994   

   Prevalence - Depressive disorders - Sex: Both - Age: All Ages (Number) (people suffering from depression)  
0                                      318435.813670                                                          
1                                      329044.773956                                                          
2                                      382544.572895                                                          
3                                      440381.507393                                                          
4                                      456916.645489

### Renombramiento
Como pudimos observar, los nombres de las columnas no siguen la convención <code>Snake Case</code>, por lo que hay que renombrar cada columna para cada una de las tablas mediante la función <code>rename</code>.

In [18]:
percentages_of_global_disorders = global_disorders.copy().rename(
    columns={
        'Entity': 'entity',
        'Code': 'code',
        'Year': 'year',
        'Schizophrenia (%)': 'schizophrenia',
        'Bipolar disorder (%)': 'bipolar_disorder',
        'Eating disorders (%)': 'eating_disorder',
        'Anxiety disorders (%)': 'anxiety',
        'Drug use disorders (%)': 'drug_addiction',
        'Depression (%)': 'depression',
        'Alcohol use disorders (%)': 'alcoholism'
    }
)

depression_rates_by_age = age_groups.copy().rename(
    columns={
        'Entity': 'entity',
        'Code': 'code',
        'Year': 'year',
        'All ages (%)': 'all',
        '10-14 years old (%)': 'from_10_to_14',
        '15-19 years old (%)': 'from_15_to_19',
        '20-24 years old (%)': 'from_20_to_24',
        '25-29 years old (%)': 'from_25_to_29',
        '30-34 years old (%)': 'from_30_to_34',
        '15-49 years old (%)': 'from_15_to_49',
        '50-69 years old (%)': 'from_50_to_69',
        '70+ years old (%)': 'above_69',
        'Age-standardized (%)': 'standardized'
    }
)

depression_rates_by_gender = gender_groups.copy().rename(
    columns = {
        'Entity': 'entity',
        'Code': 'code',
        'Year': 'year',
        'Prevalence in males (%)': 'prevalence_in_males',
        'Prevalence in females (%)': 'prevalence_in_females',
        'Population': 'population'
    }
)

depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate.copy().rename(
    columns={
        'Entity': 'entity',
        'Code': 'code',
        'Year': 'year',
        'Suicide rate (deaths per 100,000 individuals)': 'suicide_rate',
        'Depressive disorder rates (number suffering per 100,000)': 'depression_rate',
        'Population': 'population'
    }
)

depression_affected = affected.copy().rename(
    columns={
        'Entity': 'entity',
        'Code': 'code',
        'Year': 'year',
        'Prevalence - Depressive disorders - Sex: Both - Age: All Ages (Number) (people suffering from depression)': 'depression_prevalence'
    }
)

### Casting
La columna <code>year</code> de las tablas <code>gender_groups</code> y <code>depression_and_suicide_rate</code> es de tipo <code>object</code> mientras que en las demás tablas es de tipo <code>int64</code>.

1. Transformar los valores a números mediante la función <code>to_numeric</code>.
2. Filtrar los registros que no tienen valores <code>NaN</code> mediante la función <code>isna</code>.
3. Transformar los valores de <code>float64</code> a <code>int64</code> mediante la función <code>astype</code>.

In [19]:
depression_rates_by_gender['year'] = pd.to_numeric(depression_rates_by_gender['year'], errors='coerce')
depression_rates_by_gender = depression_rates_by_gender[depression_rates_by_gender['year'].isna() == False]
depression_rates_by_gender['year'] = depression_rates_by_gender['year'].astype('int64')

depression_and_suicide_rate_per_100000_individuals['year'] = pd.to_numeric(depression_and_suicide_rate_per_100000_individuals['year'], errors='coerce')
depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals[depression_and_suicide_rate_per_100000_individuals['year'].isna() == False]
depression_and_suicide_rate_per_100000_individuals['year'] = depression_and_suicide_rate_per_100000_individuals['year'].astype('int64')

### Limpieza de datos

Para comenzar, necesitamos identificar las columnas que poseen valores <code>NaN</code>, ya que estos no permiten un correcto procesamiento de los datos.

Esto se puede realizar mediante las funciones <code>isna</code> y <code>mean</code>.

Tabla de índices de desórdenes globales

In [20]:
percentages_of_global_disorders.isna().mean()

entity              0.000000
code                0.151515
year                0.000000
schizophrenia       0.000000
bipolar_disorder    0.000000
eating_disorder     0.000000
anxiety             0.000000
drug_addiction      0.000000
depression          0.000000
alcoholism          0.000000
dtype: float64

Tabla de índices de depresión por edad

In [21]:
depression_rates_by_age.isna().mean()

entity           0.000000
code             0.151515
year             0.000000
from_20_to_24    0.000000
from_10_to_14    0.000000
all              0.000000
above_69         0.000000
from_30_to_34    0.000000
from_15_to_19    0.000000
from_25_to_29    0.000000
from_50_to_69    0.000000
standardized     0.000000
from_15_to_49    0.000000
dtype: float64

Tabla de índices de depresión por género

In [22]:
depression_rates_by_gender.isna().mean()

entity                   0.000000
code                     0.034900
year                     0.000000
prevalence_in_males      0.864508
prevalence_in_females    0.864508
population               0.019356
dtype: float64

Tabla de índices de depresión y suicidio por cada 100,000 individuos

In [23]:
depression_and_suicide_rate_per_100000_individuals.isna().mean()

entity             0.000000
code               0.034900
year               0.000000
suicide_rate       0.864508
depression_rate    0.864508
population         0.019356
dtype: float64

Tabla de población afectada por la depresión

In [24]:
depression_affected.isna().mean()


entity                   0.000000
code                     0.151515
year                     0.000000
depression_prevalence    0.000000
dtype: float64

Las tablas de nuestro conjunto de datos contienen regiones del mundo, las cuales contienen información parcial y no son de gran utilidad para este análisis.

Todos estos registros pueden filtrarse mediante la función <code>isna</code> aplicada a la columna <code>code</code>, ya que esta columna no tiene valor alguno para estos registros.

In [25]:
percentages_of_global_disorders = percentages_of_global_disorders.drop(percentages_of_global_disorders[percentages_of_global_disorders['code'].isna()].index)

depression_rates_by_age = depression_rates_by_age.drop(depression_rates_by_age[depression_rates_by_age['code'].isna()].index)

depression_rates_by_gender = depression_rates_by_gender.drop(depression_rates_by_gender[depression_rates_by_gender['code'].isna()].index)

depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals.drop(depression_and_suicide_rate_per_100000_individuals[depression_and_suicide_rate_per_100000_individuals['code'].isna()].index)

depression_affected = depression_affected.drop(depression_affected[depression_affected['code'].isna()].index)

Existen registros donde la columna <code>code</code> no tiene valor alguno, por lo que se puede eliminar sin dañar la integridad de los datos.

Esto se hace mediante la función <code>drop</code>.

In [26]:
percentages_of_global_disorders = percentages_of_global_disorders.drop(columns=['code'])
depression_rates_by_age = depression_rates_by_age.drop(columns=['code'])
depression_rates_by_gender = depression_rates_by_gender.drop(columns=['code'])
depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals.drop(columns=['code'])
depression_affected = depression_affected.drop(columns=['code'])

Una vez que la columna <code>year</code> de todas las tablas es de tipo <code>int64</code>, hay que filtrar los registros comprendidos entre los años <code>1990</code> y <code>2017</code> de las tablas <code>depression_rates_by_gender</code> y <code>depression_and_suicide_rate_per_100000_individuals</code>.

In [27]:
depression_rates_by_gender = depression_rates_by_gender[depression_rates_by_gender['year'] >= 1990]
depression_rates_by_gender = depression_rates_by_gender[depression_rates_by_gender['year'] <= 2017]

depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals[depression_and_suicide_rate_per_100000_individuals['year'] >= 1990]
depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals[depression_and_suicide_rate_per_100000_individuals['year'] <= 2017]

In [28]:
print('global disorders: ', percentages_of_global_disorders.shape)
print('age groups: ', depression_rates_by_age.shape)
print('gender groups: ', depression_rates_by_gender.shape)
print('depression and suicide rate: ', depression_and_suicide_rate_per_100000_individuals.shape)
print('affected: ', depression_affected.shape)

global disorders:  (5488, 9)
age groups:  (5488, 12)
gender groups:  (6580, 5)
depression and suicide rate:  (6580, 5)
affected:  (5488, 3)


Como podemos observar, este no es el caso, ya que los dataframes de <code>depression_rates_by_gender</code> y <code>depression_and_suicide_rate_per_100000_individuals</code> aun cuentan con una mayor cantidad de filas que el resto de los datos. Para saber cualés son estos datos extras, hacemos una comparación entre los valores únicos de las columnas de <code>entity</code> de un dataframe con 5488 filas y otro con 6580 filas, con el fin de saber que entidades extras tienen estos dos datasets.

In [29]:
global_disorders_unique_list = pd.unique(percentages_of_global_disorders['entity'])
gender_groups_unique_list = pd.Series(pd.unique(depression_rates_by_gender['entity']))

countries_not_in_all_dfs = gender_groups_unique_list[gender_groups_unique_list.apply(lambda x:x not in global_disorders_unique_list)]
countries_not_in_all_dfs.head()

6                            Anguilla
10                              Aruba
25    Bonaire Sint Eustatius and Saba
29             British Virgin Islands
38                     Cayman Islands
dtype: object

Podemos observar que estas filas extras consisten en países pequeños, territorios y demás dependencias. La importancia de estos datos es de menor importancia al considerar el valor de uniformidad entre cada uno de los dataframes obtenidos. Con esto en mente eliminamos estas <code>entity</code> de los dataframes.

In [33]:
depression_rates_by_gender = depression_rates_by_gender[depression_rates_by_gender['entity'].apply(lambda x:x not in countries_not_in_all_dfs.tolist())]
depression_and_suicide_rate_per_100000_individuals = depression_and_suicide_rate_per_100000_individuals[depression_and_suicide_rate_per_100000_individuals['entity'].apply(lambda x:x not in countries_not_in_all_dfs.tolist())]

print('gender groups: ', depression_rates_by_gender.shape)
print('depression and suicide rate: ', depression_and_suicide_rate_per_100000_individuals.shape)

gender groups:  (5488, 5)
depression and suicide rate:  (5488, 5)


Con los dataframes con un número de filas uniforme, verificamos que ya no existan valores <code>NaN</code> en ninguno de los dataframes.

<code>percentages_of_global_disorders</code>

In [35]:
percentages_of_global_disorders.isna().mean()

entity              0.0
year                0.0
schizophrenia       0.0
bipolar_disorder    0.0
eating_disorder     0.0
anxiety             0.0
drug_addiction      0.0
depression          0.0
alcoholism          0.0
dtype: float64

<code>depression_rates_by_age</code>

In [36]:
depression_rates_by_age.isna().mean()

entity           0.0
year             0.0
from_20_to_24    0.0
from_10_to_14    0.0
all              0.0
above_69         0.0
from_30_to_34    0.0
from_15_to_19    0.0
from_25_to_29    0.0
from_50_to_69    0.0
standardized     0.0
from_15_to_49    0.0
dtype: float64

<code>depression_rates_by_gender</code>

In [37]:
depression_rates_by_gender.isna().mean()

entity                   0.0
year                     0.0
prevalence_in_males      0.0
prevalence_in_females    0.0
population               0.0
dtype: float64

<code>depression_and_suicide_rate_per_100000_individuals</code>

In [38]:
depression_and_suicide_rate_per_100000_individuals.isna().mean()

entity             0.0
year               0.0
suicide_rate       0.0
depression_rate    0.0
population         0.0
dtype: float64

<code>depression_affected</code>

In [39]:
depression_affected.isna().mean()

entity                   0.0
year                     0.0
depression_prevalence    0.0
dtype: float64

## Nota: Añadir continente usando la API de kaggle
https://www.kaggle.com/statchaitya/country-to-continent